File: kirun-js/src/engine/function/system/math/GenericMathFunction.ts
import { Schema } from '../../../json/schema/Schema';
import { SchemaType } from '../../../json/schema/type/SchemaType';
import { TypeUtil } from '../../../json/schema/type/TypeUtil';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { PrimitiveUtil } from '../../../util/primitive/PrimitiveUtil';
import { AbstractFunction } from '../../AbstractFunction';

const VALUE = 'value';
const VALUE1 = 'value1';
const VALUE2 = 'value2';

export class GenericMathFunction extends AbstractFunction {
    private signature: FunctionSignature;
    private parametersNumber: number;
    private mathFunction: (v1: number, v2?: number) => number;
    private readonly paramFunctions: Array<() => Map<string, Parameter>>;

    public constructor(
        functionName: string,
        mathFunction: (v1: number, v2?: number) => number,
        parametersNumber: number = 1,
        ...returnType: SchemaType[]
    ) {
        super();
        this.paramFunctions = [
            () => {
                return new Map([[VALUE, new Parameter(VALUE, Schema.ofNumber(VALUE))]]);
            },
            () => {
                return new Map([
                    [VALUE1, new Parameter(VALUE1, Schema.ofNumber(VALUE1))],
                    [VALUE2, new Parameter(VALUE2, Schema.ofNumber(VALUE2))],
                ]);
            },
        ];
        if (!returnType || !returnType.length) returnType = [SchemaType.DOUBLE];
        this.parametersNumber = parametersNumber;
        this.mathFunction = mathFunction;
        this.signature = new FunctionSignature(functionName)
            .setNamespace(Namespaces.MATH)
            .setParameters(this.paramFunctions[parametersNumber - 1]())
            .setEvents(
                new Map([
                    Event.outputEventMapEntry(
                        new Map([
                            [
                                VALUE,
                                new Schema().setType(TypeUtil.of(...returnType)).setName(VALUE),
                            ],
                        ]),
                    ) as [string, Event],
                ]),
            );
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let v1 = PrimitiveUtil.findPrimitiveNumberType(
            context.getArguments()?.get(this.parametersNumber == 1 ? VALUE : VALUE1),
        ).getT2();
        let v2;
        if (this.parametersNumber == 2)
            v2 = PrimitiveUtil.findPrimitiveNumberType(context.getArguments()?.get(VALUE2)).getT2();

        return new FunctionOutput([
            EventResult.outputOf(new Map([[VALUE, this.mathFunction.call(this, v1, v2)]])),
        ]);
    }
}

File: kirun-py/src/kirun_py/function/system/math/generic_math_function.py
from __future__ import annotations

from typing import Callable, List, Optional, TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.json.schema.type.schema_type import SchemaType
from kirun_py.json.schema.type.type_util import TypeUtil
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces
from kirun_py.util.primitive.primitive_util import PrimitiveUtil

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'value'
VALUE1 = 'value1'
VALUE2 = 'value2'


class GenericMathFunction(AbstractFunction):

    def __init__(
        self,
        function_name: str,
        math_function: Callable[..., float],
        parameters_number: int = 1,
        return_types: Optional[List[SchemaType]] = None,
    ) -> None:
        super().__init__()
        if return_types is None:
            return_types = [SchemaType.DOUBLE]

        self._parameters_number = parameters_number
        self._math_function = math_function

        param_builders = [
            lambda: {VALUE: Parameter(VALUE, Schema.of_number(VALUE))},
            lambda: {
                VALUE1: Parameter(VALUE1, Schema.of_number(VALUE1)),
                VALUE2: Parameter(VALUE2, Schema.of_number(VALUE2)),
            },
        ]

        self._signature = (
            FunctionSignature(function_name)
            .set_namespace(Namespaces.MATH)
            .set_parameters(param_builders[parameters_number - 1]())
            .set_events(
                dict([
                    Event.output_event_map_entry({
                        VALUE: Schema()
                        .set_type(TypeUtil.of(*return_types))
                        .set_name(VALUE),
                    })
                ])
            )
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        args = context.get_arguments()

        v1 = PrimitiveUtil.find_primitive_number_type(
            args.get(VALUE if self._parameters_number == 1 else VALUE1)
        ).get_t2()

        v2 = None
        if self._parameters_number == 2:
            v2 = PrimitiveUtil.find_primitive_number_type(
                args.get(VALUE2)
            ).get_t2()

        if v2 is not None:
            result = self._math_function(v1, v2)
        else:
            result = self._math_function(v1)

        return FunctionOutput([
            EventResult.output_of({VALUE: result})
        ])
